vllm/spec_decode/draft_model_runner.py

```diff
@@ -71,51 +67,148 @@ def __init__(
-                seq = seq_group_metadata.seq_data[seq_output.parent_seq_id]
+        # Update attn_metadata
+        attn_metadata = model_input.attn_metadata
+        assert isinstance(attn_metadata, FlashAttentionMetadata)
+        self._update_flash_attn_metadata(attn_metadata, num_seqs, num_queries)
 
-                token_id = seq_output.output_token
-                token_logprob = seq_output.logprobs[token_id]
+        # Update GPU tensors
+        ops.advance_step(num_seqs=num_seqs,
+                         num_queries=num_queries,
+                         block_size=self.block_size,
+                         input_tokens=model_input.input_tokens,
+                         sampled_token_ids=sampled_token_ids,
+                         input_positions=model_input.input_positions,
+                         seq_lens=attn_metadata.seq_lens_tensor,
+                         slot_mapping=attn_metadata.slot_mapping,
+                         block_tables=attn_metadata.block_tables)
 
-                seq.append_token_id(token_id, token_logprob.logprob)
-                seq.update_num_computed_tokens(1)
+        # Update sampling_metadata
+        sampling_metadata = model_input.sampling_metadata
+        self._update_sampling_metadata(sampling_metadata, num_seqs,
+                                       num_queries)
+
+        # Create new input
+        new_model_input = self._model_input_cls(
+            input_tokens=model_input.input_tokens,
+            input_positions=model_input.input_positions,
+            attn_metadata=attn_metadata,
+            seq_lens=attn_metadata.seq_lens,
+            query_lens=model_input.query_lens,
+            lora_mapping=model_input.lora_mapping,
+            lora_requests=model_input.lora_requests,
+            multi_modal_kwargs=model_input.multi_modal_kwargs,
+            sampling_metadata=model_input.sampling_metadata,
+            is_prompt=False,
+        )
 
-        return self.prepare_model_input(self.cached_seq_group_metadata_list)
+        # Ensure we skip CPU samples
+        assert new_model_input.sampling_metadata.skip_sampler_cpu_output is True
+        # We can reuse sampling tensors since every decode iteration is the same
+        new_model_input.sampling_metadata.reuse_sampling_tensors = True
+
+        if debug_advance_input:
+            print("NEW INPUT: ")
+            print("  input_tokens = {}".format(new_model_input.input_tokens))
+            print("  input_positions = {}".format(
+                new_model_input.input_positions))
+            print("  seq_lens = {}".format(new_model_input.seq_lens))
+            print("  query_lens = {}".format(new_model_input.query_lens))
+            print("  attn_metadata:")
+            print("    seq_lens_tensor: {}".format(
+                attn_metadata.seq_lens_tensor))
+            print("    slot_mapping: {}".format(attn_metadata.slot_mapping))
+            print("    block_tables: {}".format(attn_metadata.block_tables))
```

Comment:
For code consistency, I'd suggest either remove debug_advance_input or keep the flag but use logger.debug instead of print.

Reply:
Switched to logger.debug